web/src/components/vis/ScreePlot.vue

```diff
@@ -48,105 +68,93 @@ line.cutoff90 {
 }
 </style>
 
-<script>
-import { scalePoint } from 'd3-scale';
+<script lang="ts">
+import {
+  PropType, defineComponent, computed, ref, watch, onMounted,
+} from '@vue/composition-api';
+import { axisBottom, axisLeft } from 'd3-axis';
+import { scalePoint, scaleLinear } from 'd3-scale';
 import { select, event } from 'd3-selection';
 import { format } from 'd3-format';
 import { line } from 'd3-shape';
 import 'd3-transition';
 
-import { axisPlot } from './mixins/axisPlot';
+import useAxisPlot from './use/useAxisPlot';
 
-const sum = (arr) => arr.reduce((acc, x) => acc + x, 0);
+const sum = (arr: number[]) => arr.reduce((acc, x) => acc + x, 0);
 
-export default {
-  mixins: [
-    axisPlot,
-  ],
+const radius = 4;
+const margin = {
+  top: 20, right: 20, bottom: 50, left: 50,
+};
+const fadeInDuration = 500;
+const duration = 200;
 
+export default defineComponent({
   props: {
     eigenvalues: {
       required: true,
-      validator: (prop) => !prop || prop.every((v) => Number.isFinite(v) && v > 0.0),
+      type: Array as PropType<number[]>,
+      validator: (prop: number[]) => prop.every((v) => Number.isFinite(v) && v > 0.0),
     },
     numComponents: {
-      default: 10,
       type: Number,
-      validator: Number.isInteger,
+      default: 10,
     },
     showCutoffs: {
       type: Boolean,
       default: true,
     },
   },
 
-  data() {
-    return {
-      margin: {
-        top: 20,
-        right: 20,
-        bottom: 50,
-        left: 50,
-      },
-      fadeInDuration: 500,
-      duration: 200,
-      xlabel: 'Principal Component',
-      ylabel: 'Eigenvalue',
-    };
-  },
-
-  computed: {
-    eigenvaluesInternal() {
-      return this.eigenvalues || [];
-    },
-
-    numComponentsInternal() {
-      return Math.min(this.numComponents, this.eigenvaluesInternal.length);
-    },
-
-    scaleX() {
-      const {
-        numComponentsInternal,
-        dwidth,
-      } = this;
-
-      const labels = [...Array(numComponentsInternal).keys()].map((d) => d + 1);
-      return scalePoint()
+  setup(props) {
+    const svgRef = ref(document.createElement('svg'));
+    const mainRef = ref(document.createElement('div'));
+    const tooltipRef = ref(document.createElement('div'));
```

Comment:
I still don't know whether it's right to in initialize these to detached elements in order to prevent the "null" case or if the null case is there to protect me from using these before the mounted() hook (or in the event that one of these elements is inside an un-rendered `v-if`)

I definitely think this defeats the purpose of typescript, since we've effectively lied about that the actual type of the ref, but I don't know how much that matters.  I've done this too in other components, FWIW.

Reply:
I cribbed this from your work in BoxPlot, so I haven't put much thought into it.
That said, I don't see anything wrong with using this:
```
const svgRef = ref() as Ref<HTMLElement>;
```
You'll still get a runtime exception if you try to use the ref directly in `setup`, but as I understand it will be bound properly when any reactive functions are called after `setup` has completed. As long as you don't try to do any funky initializations using refs and make sure to use the `onMounted` hook to trigger the first update, everything should just work.

Regarding https://github.com/girder/viime/pull/589#discussion_r457376662, I personally thing the `onMounted(watchEffect(() => {...}));` is the least bad pattern. D3 requires the elements to be present before it can do it's thing, so it doesn't seem meaningful to be calling an update function prior to mounting.

Reply:
> You'll still get a runtime exception if you try to use the ref directly in setup

It's just not an accurate description of the type of `svgRef`

```
const svgRef = ref() as Ref<HTMLElement | undefined>;
```

This will be a compile error rather than a runtime error if you don't perform the undefined check. 

> onMounted(watchEffect(() => {...})); is the least bad pattern.

Sure, I don't mind that.

Reply:
@subdavis, do you grok the lifecycle process for setting up `ref`s? Casting the undefined values to to `Ref<HTMLElement>` has a certain simple elegance to it _if and only if_ the runtime behavior is in some way guaranteed to perform a kind of "IO monad from Haskell" magic in the background, then pave that magic over with types.

Far more likely is that this would work sort of, and then lead to issues later when, e.g., the Vue people update the type behavior for refs at some point.

I'll reiterate what I wrote in my review of #589: maybe we should file an issue asking about this and see what the Vuefolken say about it.